syntax = "proto3";
import "google/protobuf/timestamp.proto";
option java_package = "ma.enset.stubs";

service ChatServices {
  rpc connexion (stream Client) returns (stream Response);
  rpc disconnect (stream Client) returns (stream Response);
  rpc join (stream Channel) returns (stream Response);
  rpc listClient (stream Channel) returns (stream Response);
  rpc leave (stream Channel) returns (stream Response);
  rpc sendMessage (stream Message) returns (stream Response);
}

message Channel {
  string chanelName = 1;
  Client client = 2;
}

message Client{
  string name = 1;
}

message Message{
  Client from = 1;
  Channel channel = 2;
  string text = 3;
}

message Response{
  ResponseType type = 1;
  google.protobuf.Timestamp timestamp = 2;
  string message = 3;
}

enum ResponseType {
  SENT = 0;
  FAILED = 1;
  ISCONNECTED = 2;
  JOINED = 3;
  LEFT = 4;
  DISCONNECTED = 5;
  RECIEVED = 6;
  ALREADY_EXIST = 7;
}